File: 0547-number-of-provinces/0547-number-of-provinces.cpp
class Solution {
public:
    void dfs( int i, vector<vector<int>>& adj, int v, vector<int>&visited ){
        visited[i] = 1;//mark visited
        for(int neighbour = 0; neighbour<v; neighbour++){
            if(adj[i][neighbour] == 1 && !visited[neighbour]){
                dfs(neighbour, adj, v, visited);
            }
        }
    }
    int findCircleNum(vector<vector<int>>& adj) {
        int v = adj.size(); //no of vertices
        vector<int>visited(v,0);
        int provinces = 0;

        for(int i = 0; i<v; i++){
            if(!visited[i]){
            provinces++;
            dfs(i, adj, v, visited);
            }
        }

        return provinces;

    }
};

File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution:
    def dfs(self, i, visited, isConnected):
        visited[i] = 1
        for neighbour in range (0,len(isConnected)):
            if isConnected[i][neighbour] == 1 and visited[neighbour] == 0:
                self.dfs(neighbour, visited, isConnected)
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        visited = [0] * len(isConnected)
        provinces = 0
        for i in range(len(isConnected)):
            if visited[i] == 0:
                provinces += 1
                self.dfs(i, visited, isConnected)

        return provinces 